File: traversals/src/test/scala/io/shiftleft/overflowdb/testdomains/gratefuldead/ArtistDb.scala
package io.shiftleft.overflowdb.testdomains.gratefuldead

import io.shiftleft.overflowdb.{NodeLayoutInformation, NodeRef, OdbNode}
import org.apache.tinkerpop.gremlin.structure.VertexProperty
import scala.jdk.CollectionConverters._
import org.apache.tinkerpop.gremlin.util.iterator.IteratorUtils
import io.shiftleft.overflowdb.OdbNodeProperty

class ArtistDb(ref: NodeRef[ArtistDb]) extends OdbNode(ref) {
  private var _name: String = null
  def name: String = _name

  override protected def layoutInformation = Artist.layoutInformation

  override def valueMap = {
    val properties = new java.util.HashMap[String, Any]
    if (_name != null) properties.put(Artist.PropertyKeys.Name, _name)
    properties
  }
  
  override protected def specificProperties[V](key: String) = 
    key match {
      case Artist.PropertyKeys.Name if _name != null => 
        IteratorUtils.of(new OdbNodeProperty(this, key, _name.asInstanceOf[V]))
      case _ => 
        java.util.Collections.emptyIterator
    }

  override protected def updateSpecificProperty[V](cardinality: VertexProperty.Cardinality, key: String, value: V) = 
    key match {
      case Artist.PropertyKeys.Name => 
        _name = value.asInstanceOf[String]
        property(Artist.PropertyKeys.Name)
      case _ =>       
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }

  override protected def removeSpecificProperty(key: String) = 
    key match {
      case Artist.PropertyKeys.Name => 
        _name = null
      case _ =>       
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
}

File: tinkerpop3/src/test/java/overflowdb/testdomains/gratefuldead/ArtistDb.java
package overflowdb.testdomains.gratefuldead;

import overflowdb.NodeLayoutInformation;
import overflowdb.NodeRef;
import overflowdb.OdbNode;
import org.apache.tinkerpop.gremlin.structure.VertexProperty;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

public class ArtistDb extends OdbNode {
  protected ArtistDb(NodeRef ref) {
    super(ref);
  }

  private String _name;

  public String name() {
    return _name;
  }

  @Override
  public NodeLayoutInformation layoutInformation() {
    return layoutInformation;
  }

  @Override
  protected Object specificProperty2(String key) {
    if (Artist.NAME.equals(key)) {
      return _name;
    } else {
      return null;
    }
  }

  @Override
  public Map<String, Object> valueMap() {
    Map<String, Object> properties = new HashMap<>();
    if (_name != null) properties.put(Artist.NAME, _name);
    return properties;
  }

  @Override
  protected <V> VertexProperty<V> updateSpecificProperty(
      VertexProperty.Cardinality cardinality, String key, V value) {
    if (Artist.NAME.equals(key)) {
      this._name = (String) value;
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
    return property(key);
  }

  @Override
  protected void removeSpecificProperty(String key) {
    if (Artist.NAME.equals(key)) {
      this._name = null;
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
  }

  public static NodeLayoutInformation layoutInformation = new NodeLayoutInformation(
      3,
      new HashSet<>(Arrays.asList(Artist.NAME)),
      Arrays.asList(),
      Arrays.asList(SungBy.layoutInformation, WrittenBy.layoutInformation));
}
